### PR TITLE
Enabling gradient checkpointing for VAE

### PR DESCRIPTION
**Pull request by Pie31415**
PR for issue #1883

### CODE REVIEW

File: src/diffusers/models/unet_2d_condition.py
Comment (patrickvonplaten) [inline src/diffusers/models/unet_2d_condition.py:37]: The vae should not be important into the unet_2d_condition model, these two classes are unrelated

**Review by patrickvonplaten — COMMENTED** [review]
Let's try to only add gradient checkpointing for the blocks - only there we can really save memory. It doesn't help much to add it to individual layers, such as `self.conv_in` 

@williamberman could you also have a look?

**Comment by Pie31415** [comment]
> Let's try to only add gradient checkpointing for the blocks - only there we can really save memory. It doesn't help much to add it to individual layers, such as `self.conv_in`
> 
> @williamberman could you also have a look?

Got it. Thanks for the feedback.

File: src/diffusers/models/unet_2d_condition.py
Comment (Pie31415) [reply src/diffusers/models/unet_2d_condition.py:37]: Where do you suggest we add the gradient_checkpointing argument? Should this be done in `autoencoder_kl.py` or `vq_model.py`?

Comment (Pie31415) [reply src/diffusers/models/unet_2d_condition.py]: Added ability to set gradient checkpointing into `autoencoder_kl.py`

File: examples/research_projects/onnxruntime/text_to_image/train_text_to_image.py
Comment (Pie31415) [inline examples/research_projects/onnxruntime/text_to_image/train_text_to_image.py:490]: @patrickvonplaten Would this be a proper way in introducing gradient checkpointing for VAE in the training script?

File: examples/research_projects/onnxruntime/text_to_image/train_text_to_image.py
Comment (patrickvonplaten) [inline examples/research_projects/onnxruntime/text_to_image/train_text_to_image.py:398]: This should stay even with gradient checkpointing enabled.

File: examples/research_projects/onnxruntime/text_to_image/train_text_to_image.py
Comment (patrickvonplaten) [inline examples/research_projects/onnxruntime/text_to_image/train_text_to_image.py:485]: Note that gradient checkpointing doesn't necessarily mean that the model is training

File: src/diffusers/models/vae.py
Comment (patrickvonplaten) [inline src/diffusers/models/vae.py:53]: This looks nice though, but could we try to not reformat all the functions? :-)

File: src/diffusers/models/autoencoder_kl.py
Comment (patrickvonplaten) [inline src/diffusers/models/autoencoder_kl.py:226]: Let's not reformat this please :-)

**Comment by Pie31415** [comment]
> The PR looks in a much better shape now, thanks!
> 
> Could you try to revert the reformatting of unrelated functions and then let's make sure to always freeze the unet and vae of `train_text_to_image` just like before. Note that gradient checkpointing still helps even if the weights are frozen :-)
> 
> Good job! Think we're on the final stretch now

Thanks for the feedback! I really appreciate it.

Apologies on the formatting. I'm using black and it seems to really want these functions formatted this way. I might have missed this but is there a suggested linter/formatter to use?

**Comment by patrickvonplaten** [comment]
@Pie31415 could you try to use the newest black version, i.e.: https://github.com/huggingface/diffusers/blob/4ae54b37898cf2c97518038d5028645bdb816dca/setup.py#L84 - does this correspond to the version you have? :-)

File: examples/research_projects/intel_opts/textual_inversion/textual_inversion_bf16.py
Comment (Pie31415) [inline examples/research_projects/intel_opts/textual_inversion/textual_inversion_bf16.py:339]: `make style` auto formats this

File: src/diffusers/models/autoencoder_kl.py
Comment (Pie31415) [reply src/diffusers/models/autoencoder_kl.py:229]: Hmm, not sure why, but black 23.1 keeps formatting this line. I'll leave it out of the commit though

**Comment by Pie31415** [comment]
Fixed the issue with formatting. It was because of issues with line endings being CRLF.

However, some of the code formatting is still because of black (formatting on save). 
e.g. 
```python
tile = z[
  :,
  :,
  i : i + self.tile_latent_min_size,
  j : j + self.tile_latent_min_size,
 ]
```

File: examples/research_projects/onnxruntime/text_to_image/train_text_to_image.py
Comment (Pie31415) [reply examples/research_projects/onnxruntime/text_to_image/train_text_to_image.py]: How exactly is fine-tuning the VAE done in `train_text_to_image`? It seems like the issue  #1883 wants to enable gradient checkpointing specifically for training VAE independently so should we add in some sort of argument?

File: examples/research_projects/onnxruntime/text_to_image/train_text_to_image.py
Comment (patrickvonplaten) [inline examples/research_projects/onnxruntime/text_to_image/train_text_to_image.py:416]:
```suggestion
```

Think even without `requires_grad_` this should help as gradients are passed down

File: src/diffusers/models/vae.py
Comment (patrickvonplaten) [inline src/diffusers/models/vae.py:29]:
```suggestion
```

File: src/diffusers/models/vae.py
Comment (patrickvonplaten) [inline src/diffusers/models/vae.py:28]:
```suggestion
    Output of decoding method.

    Args:
```

**Review by patrickvonplaten — COMMENTED** [review]
Cool this now looks good to me! 

Last thing we should do is to add a simple test that makes sure that gradient checkpointing works. Maybe you can get some inspiration from: 
https://github.com/huggingface/diffusers/blob/07a0c1cb3fe2d12ba1a97e9455df96984b1bab78/tests/models/test_models_unet_2d_condition.py#L125

and add a similar test here:
https://github.com/huggingface/diffusers/blob/main/tests/models/test_models_vae.py

After that we can merge I think :-)

File: examples/research_projects/onnxruntime/text_to_image/train_text_to_image.py
Comment (Pie31415) [reply examples/research_projects/onnxruntime/text_to_image/train_text_to_image.py]: @patrickvonplaten 
The original issue #1883 references being able to train the VAE. How can we incorporate training specifically the VAE into this script?

Subsequently, if `requires_grad_(False)` implies a frozen VAE (e.g. no backprop is done) and the purpose of gradient checkpointing is saving memory when computing the gradient of the loss by backprop doesn't that void the entire purpose of gradient checkpointing?

Sorry for the questions but this is my first PR so I'm a trying to understand the specifics.

File: examples/research_projects/onnxruntime/text_to_image/train_text_to_image.py
Comment (williamberman) [inline examples/research_projects/onnxruntime/text_to_image/train_text_to_image.py:415]: Think can remove this as `vae.requires_grad_(False)` is called in this script :)

**Comment by williamberman** [comment]
wow super cool :) I learned how our gradient checkpointing works from this PR :grin: will approve once test is written

File: examples/research_projects/onnxruntime/text_to_image/train_text_to_image.py
Comment (Pie31415) [reply examples/research_projects/onnxruntime/text_to_image/train_text_to_image.py]: @patrickvonplaten @williamberman 
I think my above comment got lost in the resolved comment. Sorry about that.

File: examples/research_projects/onnxruntime/text_to_image/train_text_to_image.py
Comment (williamberman) [reply examples/research_projects/onnxruntime/text_to_image/train_text_to_image.py:415]: Ah yes sorry forgot model can still hold gradients if requires grad set to False. However, I don't think we're training the vae in this script and only using it for encoding images.

@Pie31415 I think that answers your original question https://github.com/huggingface/diffusers/pull/2536#discussion_r1138149998 which is that script's not meant to train the VAE so we shouldn't worry about incorporating gradient checkpointing there.

I don't think off the top of my head any of our example scripts train the vae. re: the original issue it sounds like they have their own script where they would enable it themselves

Comment (Pie31415) [reply examples/research_projects/onnxruntime/text_to_image/train_text_to_image.py:415]: @williamberman 
Okay, so just to clarify `vae.enable_gradient_checkpointing()` doesn't really do much here?

**Comment by Pie31415** [comment]
Seem to be failing CI/CD tests for MacOS. Anyway to fix this atm?

File: examples/research_projects/onnxruntime/text_to_image/train_text_to_image.py
Comment (williamberman) [reply examples/research_projects/onnxruntime/text_to_image/train_text_to_image.py:415]: Yes, that's correct! 

When computing the forward pass for the part of a network where gradients are going to have to be computed later, intermediate values have to be stored for the gradient calculation during the backward pass. This is really memory intensive! Gradient checkpointing is a trade off where we only store some of the intermdiate values and recompute the other intermediate values as needed. 

So if no gradients are calculated through that part of the network, no saving of intermediate values and thus no checkpointing is necessary.

https://arxiv.org/pdf/1910.02653.pdf This paper is more in depth but has a good section _just_ on gradient checkpointing

Comment (Pie31415) [reply examples/research_projects/onnxruntime/text_to_image/train_text_to_image.py:415]: > Yes, that's correct!
> 
> When computing the forward pass for the part of a network where gradients are going to have to be computed later, intermediate values have to be stored for the gradient calculation during the backward pass. This is really memory intensive! Gradient checkpointing is a trade off where we only store some of the intermdiate values and recompute the other intermediate values as needed.
> 
> So if no gradients are calculated through that part of the network, no saving of intermediate values and thus no checkpointing is necessary.
> 
> https://arxiv.org/pdf/1910.02653.pdf This paper is more in depth but has a good section _just_ on gradient checkpointing

So in this case is this PR for VAE gradient checkpointing necessary? If so, is the primary use when someone does have the desire to specifically train the VAE?

Comment (williamberman) [reply examples/research_projects/onnxruntime/text_to_image/train_text_to_image.py:415]: Yep PR adding it still helpful! And yes that's the primary use case :)

Comment (Pie31415) [reply examples/research_projects/onnxruntime/text_to_image/train_text_to_image.py:415]: Perfect! Thank you for your insight :)

**Review by williamberman — APPROVED** [review]

**Comment by patrickvonplaten** [comment]
Good job @Pie31415

**Pull request merged**
